### ISSUE TITLE
Changed callbacks firing before `connect`

### ISSUE THREAD

**Issue by gahia** [hotwired/stimulus#490]
Hi!

Using Changed callbacks I've found these are always executed before the 'connect' method. Is this an intended behaviour?

```js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {

  static values = { index: Number }

  connect() {
    console.debug(this.element.id, "connect...")
  }

  indexValueChanged() {
    console.debug(this.element.id, "indexValueChanged...")
  }
}
```
```html
<div
  id="without_index_value"
  data-controller="test"
></div>

<div
  id="with_index_value"
  data-controller="test"
  data-test-index-value="1"
></div>
```
```
without_index_value indexValueChanged... 
without_index_value connect... 
with_index_value indexValueChanged... 
with_index_value connect...
```

Using Stimulus 3.0.1

**Comment by seanpdoyle** [comment]
@gahia yes, this is the intended behavior:

> Stimulus invokes each change callback after the controller is initialized and again any time its associated data attribute changes. This includes changes as a result of assignment to the value’s setter.

You can learn more about [Change Callbacks](https://stimulus.hotwired.dev/reference/values#change-callbacks) on the Stimulus documentation site.

**Comment by gahia** [comment]
Thanks @seanpdoyle. I've read that quote in the docs, but was not clear for me what was the exact order in the lifecycle as "after the controller is initialized" does not necessarily implies the order against the other methods/callbacks in the lifecycle.
 
Would it worth to add the change callbacks in the [Lifecycle Callbacks](https://stimulus.hotwired.dev/reference/lifecycle-callbacks)?

**Comment by wynksaiddestroy** [comment]
> @gahia yes, this is the intended behavior:
> 
> > Stimulus invokes each change callback after the controller is initialized and again any time its associated data attribute changes. This includes changes as a result of assignment to the value’s setter.
> 
> You can learn more about [Change Callbacks](https://stimulus.hotwired.dev/reference/values#change-callbacks) on the Stimulus documentation site.

The change callback is called even before the controller is initialized. So there is no chance to setup anything. I cobbled together a small example:

```html
<!doctype html>
<html>
<head>
  <meta charset="utf-8">
  <script type="module">
    import { Application, Controller } from "https://unpkg.com/@hotwired/stimulus/dist/stimulus.js"
    window.Stimulus = Application.start()

    Stimulus.register("hello", class extends Controller {
      static values = { "name": { type: String, default: "John" } }

      initialize() {
      	console.log("initialize called");
      }

      nameValueChanged() {
      	console.log("nameValueChanged called");
      }
    })
  </script>
</head>
<body>
  <div data-controller="hello"></div>
</body>
</html>
```

**Comment by seanpdoyle** [comment]
You're right, the ValueObserver that backs the controller context calls `this.invokeChangedCallbacksForDefaultValues()` in its constructor, instead of its `start()` method when a `default:` option is provided.

This issue should be re-opened.

**Comment by wynksaiddestroy** [comment]
Thank you vey much @seanpdoyle ❤️

**Comment by marcoroth** [comment]
Hey @christophehenry, it might sound imitating to upgrade from 2.x to 3.x but   technically it should be as easy as bumping your `stimulus` version in your `package.json` from 2.x to 3.x.

You don't need to re-rewrite anything as everything should still be compatible, to ease the upgrade process.

For later it's recommended that you migrate from `stimulus` to `@hotwired/stimulus`. It was only released as new major version because of renaming of the library. 

But everything else is bascially also available under the old library name.

**Comment by christophehenry** [comment]
@marcoroth Can't do. Because we don't use any build system and we need to keep IE11 compatibility, unhopefully :pensive:

**Comment by marcoroth** [comment]
@christophehenry FWIW, the `@stimulus/polyfills` is still around is technically still compatible with v3. I haven't tried it myself but I don't see why it should work. Maybe it's worth a shot.

**Comment by marcoroth** [comment]
We should be able to close this since this was resolved via #499
